types/block.go

```diff
@@ -68,3 +73,157 @@ type Signature []byte
 type IntermediateStateRoots struct {
 	RawRootsList [][]byte
 }
+
+func (h *Header) ToProto() *pb.Header {
+	return &pb.Header{
+		Version: &pb.Version{
+			Block: h.Version.Block,
+			App:   h.Version.App,
+		},
+		NamespaceId:     h.NamespaceID[:],
+		Height:          h.Height,
+		Time:            h.Time,
+		LastHeaderHash:  h.LastHeaderHash[:],
+		LastCommitHash:  h.LastCommitHash[:],
+		DataHash:        h.DataHash[:],
+		ConsensusHash:   h.ConsensusHash[:],
+		AppHash:         h.AppHash[:],
+		LastResultsHash: h.LastResultsHash[:],
+		ProposerAddress: h.ProposerAddress[:],
+	}
+}
+
+func (h *Header) FromProto(other *pb.Header) error {
+	h.Version.Block = other.Version.Block
+	h.Version.App = other.Version.App
+	if !safeCopy(h.NamespaceID[:], other.NamespaceId) {
+		return ErrInvalidNamespaceId
+	}
+	h.Height = other.Height
+	h.Time = other.Time
+	if !safeCopy(h.LastHeaderHash[:], other.LastHeaderHash) {
+		return ErrInvalidLastHeaderHash
+	}
+	if !safeCopy(h.LastCommitHash[:], other.LastCommitHash) {
+		return ErrInvalidLastCommitHash
+	}
+	if !safeCopy(h.DataHash[:], other.DataHash) {
+		return ErrInvalidDataHash
+	}
+	if !safeCopy(h.ConsensusHash[:], other.ConsensusHash) {
+		return ErrInvalidConsensusHash
+	}
+	if !safeCopy(h.AppHash[:], other.AppHash) {
+		return ErrInvalidAppHash
+	}
+	if !safeCopy(h.LastResultsHash[:], other.LastResultsHash) {
+		return ErrInvalidLastResultsHash
+	}
+	if !safeCopy(h.ProposerAddress[:], other.ProposerAddress) {
+		return ErrInvalidProposerAddress
+	}
+
+	return nil
+}
+
+func safeCopy(dst, src []byte) bool {
+	n := copy(dst, src)
+	return n == len(dst)
```

Comment:
shouldn't this return an error in the case `n != len(dst)` instead? Otherwise, it's not so clear what the bool means. On the other hand, only an unexported helper and `if !safeCopy` is easy to understand and read.

Reply:
I agree that returning an error would be more idiomatic. A semantic concern is: isn't `n` guaranteed to be equal to `len(dst)` _after_ the copy?

Reply:
> Copy returns the number of elements copied, which will be the minimum of len(src) and len(dst).

 https://golang.org/pkg/builtin/#copy

Reply:
Ah I misunderstood that line, so if `len(dst) > len(src)`, then only `len(src)` bytes are copied, and the final `len(dst)` will remain the same as the initial length. For some reason I understood that the destination slice would get completely clobbered.

Reply:
Even if error is returned, it will be ignored, as other, more detailed one will be returned from calling method.
And in my opinion `if !safeCopy(...) { return someError; }` is more expressive than `if err := safeCopy(...); err != nil { return someErr; }`

I'll add godoc, to document the behavior of this function.

Reply:
I agree with @tzdybal that for this func keeping bool is nicer. Furthermore, returning bool instead of error is idiomatic as well, specifically for such little helper funcs.

Reply:
Actually I changed the code to be more explicit and to actually handle all cases (len(src) > len(dst), len(src) < len(src)). Now it will return true only if the size of both slices is the same, so we can use it for actual validation of fixed-size arrays.